src/Framework/StringBuilderCache.cs

```diff
@@ -51,27 +55,44 @@ public static StringBuilder Acquire(int capacity = 16 /*StringBuilder.DefaultCap
             if (capacity <= MAX_BUILDER_SIZE)
             {
                 StringBuilder sb = StringBuilderCache.t_cachedInstance;
+                StringBuilderCache.t_cachedInstance = null;
                 if (sb != null)
                 {
                     // Avoid StringBuilder block fragmentation by getting a new StringBuilder
                     // when the requested size is larger than the current capacity
                     if (capacity <= sb.Capacity)
                     {
-                        StringBuilderCache.t_cachedInstance = null;
                         sb.Length = 0; // Equivalent of sb.Clear() that works on .Net 3.5
+#if DEBUG && !CLR2COMPATIBILITY && !MICROSOFT_BUILD_ENGINE_OM_UNITTESTS
+                        MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: sb.GetHashCode(), newCapacity: capacity, oldCapacity: sb.Capacity, type: "sbc-hit");
+#endif
                         return sb;
                     }
                 }
             }
-            return new StringBuilder(capacity);
+
+            StringBuilder stringBuilder = new StringBuilder(capacity);
+#if DEBUG && !CLR2COMPATIBILITY && !MICROSOFT_BUILD_ENGINE_OM_UNITTESTS
+            MSBuildEventSource.Log.ReusableStringBuilderFactoryStart(hash: stringBuilder.GetHashCode(), newCapacity: capacity, oldCapacity: stringBuilder.Capacity, type: "sbc-miss");
+#endif
+            return stringBuilder;
         }
 
         public static void Release(StringBuilder sb)
         {
             if (sb.Capacity <= MAX_BUILDER_SIZE)
             {
+#if ASSERT_BALANCE
```

Comment:
Should we do this for all debug builds?

Reply:
I had it that way but then it was failing at unit test in CI. I originally though, "well, thats because xunit runs it in multithread" but that was flawed explanation as it is ThreadStatic and concurrency shall not affect it.
I have locally reproduced it and found bug in code which acquires two SB at once. I will fix it, and replace ASSERT_BALANCE by DEBUG as this really identify wrong usage patterns.

Reply:
done